lib/Analysis/DependenceAnalysis.cpp

```diff
@@ -131,17 +131,35 @@ circt::analysis::MemoryDependenceAnalysis::MemoryDependenceAnalysis(
   // Collect affine loops grouped by nesting depth.
   std::vector<SmallVector<AffineForOp, 2>> depthToLoops;
   mlir::affine::gatherLoops(funcOp, depthToLoops);
+  depthToLoops.erase(depthToLoops.begin());
 
   // Collect load and store operations to check.
-  SmallVector<Operation *> memoryOps;
-  funcOp.walk([&](Operation *op) {
-    if (isa<AffineReadOpInterface, AffineWriteOpInterface>(op))
-      memoryOps.push_back(op);
-  });
-
-  // For each depth, check memref accesses.
-  for (unsigned depth = 1, e = depthToLoops.size(); depth <= e; ++depth)
-    checkMemrefDependence(memoryOps, depth, results);
+  auto rootReads = funcOp.getOps<AffineReadOpInterface>();
+  auto rootWrites = funcOp.getOps<AffineWriteOpInterface>();
+  SmallVector<Operation *> rootMemoryOps(rootReads.begin(), rootReads.end());
+  rootMemoryOps.append(rootWrites.begin(), rootWrites.end());
+
+  // Need to loop over each root loop to preserve loop nest structure
+  for (auto forOp : funcOp.getOps<AffineForOp>()) {
+    SmallVector<Operation *> memoryOps = rootMemoryOps;
+    forOp.walk([&](Operation *op) {
```

Comment:
Doing the check memref dependence in a loop makes sense, and this walk makes sense, but do we need the previous rootReads and rootWrites if we also do this? The getOps call doesn't descend into regions, so is it required to collect memory operations outside of any loops?

Reply:
>  is it required to collect memory operations outside of any loops?

I was just confused whether memory ops outside of a loop nest are needed are not for dependence analysis. The removed lines at L136 _do_ include those ops, so I was trying to imitate that behavior.

Reply:
I see, we did previously consider all the affine memory ops potentially floating around in the top level, so I guess it makes sense to keep this.

Reply:
Separately, can you please add/duplicate the previous comment to say something like "Collect load and store operations inside the current ForOp".